### ISSUE TITLE
Support un-shuffleable tracks for DJ mixes / classical pieces

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#703]
Original [issue 703](https://code.google.com/p/quodlibet/issues/detail?id=703) created by nick.boultbee on 2011-04-03T20:27:50.000Z:

From wiki post (http://code.google.com/p/quodlibet/wiki/Contact): 

Is there a way to set up Quod Libet so that it plays all the movements of a classical music piece in sequence? I prefer to use shuffle, rather than bother with creating playlists. But when QL happens to pick a track in its shuffle that is the first movement of a piece, I'd like it to play it, then play all the other movements of the same piece, in sequence, then go back to shuffle. And if QL picks a track in its shuffle that is the second, third, fourth, etc. movement in its shuffle, I'd it to detect that they are non-initial movements and skip over them. I know I can create tags to keep track of the movements. I just don't know how to get QL to use them the way I describe.

**Comment by archy3** [comment]
This too has been an issue for me.

One solution I can think of is in the pane browser, allow for random item selection in the rightmost pane upon completion of the Song List.

So if I group all movements in a piece together under the grouping tag, and I want to shuffle by grouping, I could set my rightmost pane to <grouping|<grouping><album>|<album><discnumber><tracknumber>> and then shuffle that pane.

This would be very flexible and extendible. The previous panes would allow for easy filtering, and users could define groups however they want. It probably wouldn't be to hard to extend this to more than just the rightmost pane, so one day someone could shuffle by the Grouping pane and another day shuffle by the Album pane (all panes to its right would be set to all) without first removing the grouping pane.

The only downside is that having a pane like <grouping|<grouping><album>|<album><discnumber><tracknumber>> would add a lot of clutter and probably would not be used to actually drill down the library but just for this randomization purpose. Perhaps "randomization only" panes could be made invisible though it would make the Song List look odd given the visibly selected pane items (that probably wouldn't be a good fix, just speaking out loud here).

The above is the simplest solution I can think of. It could almost be already implemented by swapping the Grouping tag with the Album tag and using the Random Album Plugin [but then I wouldn't be able to shuffle by album anymore :(].

The ideal solution would be to have the shuffle function have a user defined grouping mechanism. The Song List would display all the songs that are normal for that browser, but the shuffle function would pretend there is a pane with the grouping specification (that only contains songs from the Song List). It would randomly select an item from that pane and play all songs from it, then randomly select another pane item -- but to the users eyes it's just shuffling around in the Song List (but by grouping of course). The Song List is never abridged to the just the specific group of songs. This would act like the normal shuffle function but with grouping, and would work in all browsers.

It's like the invisible pane idea, but the Song List wouldn't be abridged.

**Comment by archy3** [comment]
I have created a shell script that almost fixes this issue. Say all movements of the same work have an identical 'grouping' tag. We could shuffle by that tag, just as we can shuffle by albums, with `quodlibet --random=grouping`. This script issues “quodlibet --random=grouping”, plays the song list, detect when the song list has finished, and issues “quodlibet --random=grouping” again.

**For Quod Libet 3.2.2:**
```
#!/bin/sh

## Quod Libet 3.2.2 shuffle by grouping Bourne shell script.

## This may not work in all browsers. If it doesn't work in your preferred
## browser, switch to the search browser.

## Defines grouping tag as first argument, else defaults to 'grouping'.
tag=${1:-'grouping'}

while :; do

  if [ "$(quodlibet --status)" = "" ] && sleep 1 &&
     [ "$(quodlibet --status)" = "" ]; then

    quodlibet --random=$tag &&
    sleep 4 && ## See note
    quodlibet --next &&
    sleep 2

  fi

done

## NOTE: The value '4' will need to be larger for larger libraries.
```
**For Quod Libet 3.8.1:**
```
#!/bin/sh

## Quod Libet 3.8.1 shuffle by grouping Bourne shell script.

## This may not work in all browsers. If it doesn't work in your preferred
## browser, switch to the search browser.

## Defines grouping tag as first argument, else defaults to 'grouping'.
tag=${1:-'grouping'}

control=$HOME/.quodlibet/control

while :; do

  if [ "$(quodlibet --status | tail -c 6)" = "0.000" ] && sleep 1 &&
     [ "$(quodlibet --status | tail -c 6)" = "0.000" ]; then

    test -p $control && echo "random $tag" > $control || exit 1

    sleep 4 && ## See note
    test -p $control && echo "next" > $control || exit 2

    sleep 2

  fi

done

## For some reason "quodlibet --random=tag" wasn't working for me with 3.8.1,
## hence the use of '$HOME/.quodlibet/control'.

## NOTE: The value '4' will need to be larger for larger libraries.
```
**Caveats:**
Say a group of movements from the album “Sammy Conducts Beethoven” have a grouping tag of “Symphony No. 5” and another group of movements from the album “Samantha Conducts Beethoven” also have a grouping tag of “Symphony No. 5”. Those two different works (well different recordings) will be grouped together because they share the same grouping tag.
The solution around this should be to use tied tags, instead of `quodlibet --random=grouping`, `quodlibet --random=~album~grouping` should be used. However `quodlibet --random=~tied~tag` behaves oddly. See issue #2344.

You probably would not assign a grouping tag to a single movement piece, but that piece will be excluded from this shuffle if it does not have a grouping tag.
Solution 1: Copy the track title to the grouping tag of all single movement works.
Solution 2: Allow OR options in tags like is allowed for displays, something like `quodlibet --random=<grouping|<grouping>|<~filename>>` (this is not currently supported but just giving ideas to people. It would have the effect of selecting a random item from a pane set by `<grouping|<grouping>|<~filename>>`).

In the Pane browser, selecting pane items will not act as a filter. So if you only want to listen to Concertos and Symphonies, selecting Concerto and Symphony from the `<genre>` pane will not achieve that.
Solution: The search bar can be used to act as a filter, `|(genre=Concerto, genre=Symphony)` would achieve the above.

**Comment by SpaceAmoeba** [comment]
Quodlibet is by far the best music player I have used. However, this is the one missing feature that causes frustration. I'd really like to have a shuffle-by-grouping option. 
I looked at the code for RandomAlbum (which allows for shuffling by album). I'm not very familiar with python but it looks like it relies on quodlibet keeping track of album lists, which it probably doesn't do for groupings.
What is necessary to have this functionality?
